### DISCUSSION TITLE
Bfloat16

### DISCUSSION THREAD

**Discussion by npiroozan** [deepmodeling/deepmd-kit#1933] [Q&A] [answered]
Hello,

Does DeePMD support BFloat16?

Warm Regards and thank you.

**Comment by njzjz** [comment, marked as answer]
New types can be added here:
https://github.com/deepmodeling/deepmd-kit/blob/c39339247a8a5a254fa7509be898771ab00498e9/deepmd/common.py#L41-L45

Contribution is welcome if you have found bfloat 16 faster. We do not have a hardware that supports bfloat 16.

**Reply by npiroozan** [reply]
Thank you.  Yes, I will get this implemented.  Since NumPy is used extensively, this does present a slight complication.  From what I have seen, NumPy does not support BF16 at the moment.  Are you aware of any suitable patch for this?

Thank you!

**Reply by njzjz** [reply]
You can input FP32 data and cast to BF16 in the neural network, which is implemented using TensorFlow. This is what [`precision`](https://docs.deepmodeling.com/projects/deepmd/en/v2.1.5/train/train-input.html#argument:model/descriptor[se_e2_a]/precision) parameter does.

**Reply by npiroozan** [reply]
Thank you very much for the direction on this.  Would it be possible to collaborate?  Would this implementation be a significant change to the code?

**Reply by njzjz** [reply]
See #1970.

**Reply by npiroozan** [reply]
Outstanding, thank you very much.  To implement this with LAMMPS, is anything additional required?  We can input data from LAMMPS in FP32.

**Reply by njzjz** [reply]
No, the graph will be processed by TensorFlow.

**Reply by npiroozan** [reply]
I'll let you know what type of results are observed for training, inference, and DP+LMP.  It should be very significant indeed.  Thank you very much.

**Reply by npiroozan** [reply]
Significant increase across training, inference, and DP+LMP.  I utilized mixed precision for training.

However, I am curious if it is possible to use bfloat16 for the embedding net and fitting net as well?

**Reply by npiroozan** [reply]
Please ignore the above question.  Resolved.
